src/charm.py

```diff
@@ -0,0 +1,323 @@
+            db_port=None,
+        )
+
+        self.db = pgsql.PostgreSQLClient(self, "db")  # pylint: disable=C0103
+        self.framework.observe(
+              self.db.on.database_relation_joined,
+              self._on_database_relation_joined
+          )
+        self.framework.observe(
+              self.db.on.master_changed,
+              self._on_master_changed
+          )
+
+    def _on_database_relation_joined(
+         self, event: pgsql.DatabaseRelationJoinedEvent  # type: ignore
+    ) -> None:
+        """Handle db-relation-joined.
+
+        Args:
+            event: Event triggering the database relation joined handler.
+        """
+        if self.model.unit.is_leader():
+            # Provide requirements to the PostgreSQL server.
+            event.database = DATABASE_NAME
+            event.extensions = ["hstore:public", "pg_trgm:public"]
+        elif event.database != DATABASE_NAME:
+            # Leader has not yet set requirements. Defer, in case this unit
+            # becomes leader and needs to perform that operation.
+            event.defer()
+
+    # pgsql.MasterChangedEvent is actually defined
+    def _on_master_changed(self, event: pgsql.MasterChangedEvent) -> None:
+        """Handle changes in the primary database unit.
+
+        Args:
+            event: Event triggering the database master changed handler.
+        """
+        if event.master is None:
+            self._stored.db_name = None
+            self._stored.db_user = None
+            self._stored.db_password = None
+            self._stored.db_host = None
+            self._stored.db_port = None
+        else:
+            self._stored.db_name = event.master.dbname
+            self._stored.db_user = event.master.user
+            self._stored.db_password = event.master.password
+            self._stored.db_host = event.master.host
+            self._stored.db_port = event.master.port
+
+        self._on_config_changed(event)
+
+    def _on_install(self, _):
+        """Generate a config template to be rendered later."""
+        self._run_generate_synapse()
+
+    @property
+    def _external_hostname(self):
+        """Return the external hostname passed to ingress via the relation."""
+        # It is recommended to default to `self.app.name` so that the external
+        # hostname will correspond to the deployed application name in the
+        # model, but allow it to be set to something specific via config.
+        return self.config["server_name"] or self.app.name
+
+    def _container(self):
+        """Get the Synapse workload container.
+
+        Returns:
+            The pebble instance of the Synapse container.
+        """
+        return self.unit.get_container(self._CONTAINER_NAME)
+
+    def _populate_synapse_env_settings(self):
+        """Populate env settings to pass around."""
+        if os.getenv("SYNAPSE_SERVER_NAME") is None:
+            os.environ["SYNAPSE_SERVER_NAME"] = self.config["server_name"]
```

Comment:
Is this intending to manipulate the environment of the charm container?

Reply:
Yes, same approach as in wordpress charm, this is keeping a list of env variables that are fed to pebble.